virtcontainers/fc.go

```diff
@@ -546,6 +555,85 @@ func (fc *firecracker) fcSetVMBaseConfig(mem int64, vcpus int64, htEnabled bool)
 	return err
 }
 
+func (fc *firecracker) fcSetLogger() error {
+	span, _ := fc.trace("fcSetLogger")
+	defer span.Finish()
+
+	fcLogLevel := "Error"
+
+	var err error
+	fcLogFifoFile := filepath.Join(fc.vmPath, fcLogFifo)
+	fc.logFifo, err = fifo.OpenFifo(context.Background(), fcLogFifoFile, syscall.O_CREAT|syscall.O_RDONLY|syscall.O_NONBLOCK, 0)
+	if err != nil {
+		return fmt.Errorf("Failed to open/create firecracker log fifo file")
+	}
+
+	jailedLogFifo, err := fc.fcJailResource(fcLogFifoFile, fcLogFifo)
+	if err != nil {
+		fc.Logger().WithField("set logger failed", err).Error()
+		return err
+	}
+
+	// listen on log fifo file and transfer error info
+	go func() {
+		scanner := bufio.NewScanner(fc.logFifo)
+		for scanner.Scan() {
+			fc.Logger().WithFields(logrus.Fields{
+				"level":    "log",
+				"contents": scanner.Text()}).Error("firecracker failed")
+		}
+
+		if err := scanner.Err(); err != nil {
+			fc.Logger().WithError(err).Errorf("Failed reading firecracker logs from fifo file")
+		}
+
+		if err := fc.logFifo.Close(); err != nil {
+			fc.Logger().WithError(err).Errorf("Failed closing firecracker log fifo file")
+		}
+	}()
+
+	fcMetricsFifoFile := filepath.Join(fc.vmPath, fcMetricsFifo)
+	fc.metricsFifo, err = fifo.OpenFifo(context.Background(), fcMetricsFifoFile, syscall.O_CREAT|syscall.O_RDONLY|syscall.O_NONBLOCK, 0)
+	if err != nil {
+		return fmt.Errorf("Failed to open/create firecracker metrics fifo file")
+	}
+
+	jailedMetricsFifo, err := fc.fcJailResource(fcMetricsFifoFile, fcMetricsFifo)
+	if err != nil {
+		fc.Logger().WithField("set logger failed", err).Error()
+		return err
+	}
+
+	// listen on metrics file and transfer error info
+	go func() {
+		scanner := bufio.NewScanner(fc.metricsFifo)
+		for scanner.Scan() {
+			fc.Logger().WithFields(logrus.Fields{
+				"level":    "metrics",
+				"contents": scanner.Text()}).Error("firecracker failed")
+		}
+
+		if err := scanner.Err(); err != nil {
+			fc.Logger().WithError(err).Error("Failed reading firecracker metrics from fifo file")
+		}
+
+		if err := fc.metricsFifo.Close(); err != nil {
+			fc.Logger().WithError(err).Error("Failed closing firecracker metrics fifo file")
+		}
+	}()
```

Comment:
I think you can factorize a lot of code here, how about creating a new function to open and read the pipe and call it with different parameters depending on the pipe type

Reply:
Hi @devimc I have already used the API from [containerd/fifo](https://github.com/containerd/fifo) to open, read,..., close pipe files. 
And you're right, reading from l`ogs.fifo` and `metrics.fifo` share a lot same logic and I should create one new function to summarize. ;)